esm-loader.mjs

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2022 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import newrelic from './index.js'
+import shimmer from './lib/shimmer.js'
+
+/**
+ * Hook chain responsible for resolving a file URL for a given module specifier
+ *
+ * Our loader has to be the last user-supplied loader if chaining is happening,
+ * as we rely on `nextResolve` being the default Node.js resolve hook to get our URL
+ *
+ * Docs: https://nodejs.org/api/esm.html#resolvespecifier-context-nextresolve
+ *
+ * @param {string} specifier
+ *        String identifier in an import statement or import() expression
+ * @param {object} context
+ *        Metadata about the specifier, including url of the parent module and any import assertions
+ *        Optional argument that only needs to be passed when changed
+ * @param {Function} nextResolve
+ *        The Node.js default resolve hook
+ * @returns {Promise} Promise object representing the resolution of a given specifier
+ */
+export async function resolve(specifier, context, nextResolve) {
+  if (!newrelic.shim) {
+    return nextResolve(specifier)
+  }
+
+  const instrumentationApi = newrelic.shim
+  const logger = instrumentationApi.logger.child({ component: 'esm-loader' })
+
+  const { url, format } = await nextResolve(specifier)
+  const instrumentationName = shimmer.getInstrumentationNameFromModuleName(specifier)
+  const instrumentationDefinition = shimmer.registeredInstrumentations[instrumentationName]
+
+  if (instrumentationDefinition) {
+    logger.debug(`Instrumentation exists for ${specifier}`)
+
+    if (format === 'commonjs') {
+      // ES Modules translate import statements into fully qualified filepaths, so we create a copy of our instrumentation under this filepath
+      const instrumentationDefinitionCopy = Object.assign({}, instrumentationDefinition)
+
+      // Stripping the prefix is necessary because the code downstream gets this url without it
+      instrumentationDefinitionCopy.moduleName = url.replace('file://', '')
```

Comment:
To make sure it doesn't get lost, is there a specific story for tweaking/updating/fixing the automatically generated metrics for the instrumentation for these use cases? If not, you'll want a story (or to fix in this PR). 

Without updating that code, instead of generating metrics like: 
* `Supportability/Features/Instrumentation/OnRequire/express`
* `Supportability/Features/Instrumentation/OnRequire/express/Version/4`

You'll end up with metrics like:
* `Supportability/Features/Instrumentation/OnRequire//Users/theuser/some/path/to/app/node_modules/express/index.js`
* `Supportability/Features/Instrumentation/OnRequire//Users/theuser/some/path/to/app/node_modules/express/index.js/Version/4\`

Reply:
Ah nevermind, I found it mentioned in https://issues.newrelic.com/browse/NEWRELIC-3120. Leaving for note/reminder I checked this but I'll quickly resolve it.